src/org/sosy_lab/java_smt/example/NQueens.java

```diff
@@ -147,69 +147,114 @@ BooleanFormula[][] getSymbols() {
     public List<BooleanFormula> getRules(BooleanFormula[][] symbols, SolverContext context) {
         List<BooleanFormula> rules = new ArrayList<>();
         int n = symbols.length;
 
-        // At least one queen per row
-        for (BooleanFormula[] pSymbol : symbols) {
-            List<BooleanFormula> clause = new ArrayList<>(Arrays.asList(pSymbol).subList(0, n));
+        /*Rule 1: At least one queen per row,
+        or we can say make sure that there are N Queens on the board
+        *
+        */
+        for (BooleanFormula[] rowSymbols : symbols) {
+            List<BooleanFormula> clause = new ArrayList<>(Arrays.asList(rowSymbols).subList(0, n));
             rules.add(this.bmgr.or(clause));
         }
 
-        // At most one queen per row
-        for (BooleanFormula[] pSymbol : symbols) {
+        /*Rule 2: Add constraints to ensure that at most one queen is placed in each row.
+         * For n=4:
+         *   0123
+         * 0 ----
+         * 1 ----
+         * 2 ----
+         * 3 ----
+         * We add a negation of the conjunction of all possible pairs of variables in each row.
+         */
+        for (BooleanFormula[] rowSymbol : symbols) {
             for (int j1 = 0; j1 < n; j1++) {
                 for (int j2 = j1 + 1; j2 < n; j2++) {
-                    rules.add(bmgr.not(bmgr.and(pSymbol[j1], pSymbol[j2])));
+                    rules.add(bmgr.not(bmgr.and(rowSymbol[j1], rowSymbol[j2])));
                 }
             }
         }
-        // At most one queen per column
+        /* Rule 3: Add constraints to ensure that at most one queen is placed in each column.
+         * For n=4:
+         *   0123
+         * 0 ||||
+         * 1 ||||
+         * 2 ||||
+         * 3 ||||
+         * We add a negation of the conjunction of all possible pairs of variables in each column.
+         */
         for (int j = 0; j < n; j++) {
             for (int i1 = 0; i1 < n; i1++) {
                 for (int i2 = i1 + 1; i2 < n; i2++) {
                     rules.add(bmgr.not(bmgr.and(symbols[i1][j], symbols[i2][j])));
                 }
             }
         }
-        // At most one queen per diagonal
+        /* Rule 4: At most one queen per diagonal
+         transform the field (=symbols) from square shape into a (downwards/upwards directed)
+         rhombus that is embedded in a rectangle (=DownwardDiagonal/UpwardDiagonal)
+
+         For example for N=4 from this square:
+         0123
+         0 xxxx
+         1 xxxx
+         2 xxxx
+         3 xxxx
+
+         to this rhombus/rectangle:
+         0123
+         0 x---
+         1 xx--
+         2 xxx-
+         3 xxxx
+         4 -xxx
+         5 --xx
+         6 ---x*/
         int numDiagonals = 2 * n - 1;
-        BooleanFormula[][] diagonals1 = new BooleanFormula[numDiagonals][n];
-        BooleanFormula[][] diagonals2 = new BooleanFormula[numDiagonals][n];
+        BooleanFormula[][] DownwardDiagonal = new BooleanFormula[numDiagonals][n];
```

Comment:
Variable names should be written in camelCase. they should start with lowercase letter.